src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -66,20 +66,64 @@ private void BeforeExecuteCommand(TypeCharCommandArgs args, CommandExecutionCont
             }
 
             var caret = caretOpt.Value;
-
             var document = caret.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
             if (document == null)
             {
                 return;
             }
 
-            // on the UI thread
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
             var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+
+            var currentNode = GetStartingNode(document, root, caret);
+            MoveCaretToSemicolonPosition(args, document, root, caret, syntaxFacts, currentNode, isInsideDelimiters: false);
+        }
+
+        private void MoveCaretToSemicolonPosition(TypeCharCommandArgs args, Document document, SyntaxNode root, SnapshotPoint caret, ISyntaxFactsService syntaxFacts, SyntaxNode currentNode, bool isInsideDelimiters)
+        {
+            if (currentNode == null ||
+                IsInAStringOrCharacter(currentNode, caret))
+            {
+                // Don't complete statement.  Return without moving the caret.
+                return;
+            }
+
+            if (currentNode.IsKind(SyntaxKind.ArgumentList, SyntaxKind.ArrayRankSpecifier, SyntaxKind.BracketedArgumentList, SyntaxKind.ParenthesizedExpression))
+            {
+                // make sure the closing delimiter exists
+                if (RequiredDelimiterIsMissing(currentNode))
+                {
+                    return;
+                }
+                // set caret to just outside the delimited span and analyze again
+                var newCaretPosition = currentNode.Span.End;
+                var newCaret = args.SubjectBuffer.CurrentSnapshot.GetPoint(newCaretPosition);
+                currentNode = GetStartingNode(document, root, newCaret);
+                MoveCaretToSemicolonPosition(args, document, root, newCaret, syntaxFacts, currentNode, isInsideDelimiters: true);
+                return;
+            }
+            else if (syntaxFacts.IsStatement(currentNode) || currentNode.IsKind(SyntaxKind.FieldDeclaration))
+            {
+                MoveCaretToFinalPositionInStatement(currentNode, args, caret, isInsideDelimiters);
+                return;
+            }
+            else
+            {
+                // keep caret the same, but continue analyzing with the parent of the current node
+                currentNode = currentNode.Parent;
+                MoveCaretToSemicolonPosition(args, document, root, caret, syntaxFacts, currentNode, isInsideDelimiters);
+                return;
+            }
+        }
+
+        private SyntaxNode GetStartingNode(Document document, SyntaxNode root, SnapshotPoint caret)
+        {
+            // on the UI thread
```

Comment:
Since this function doesn't change threads, maybe move the comment for the UI thread to be a function comment? Or explicitly call it `GetStartingNode_UIThread` to let callers know they need to be on the UI thread.

Reply:
Doc comment added.

Reply:
@chborl I don't see the doc comment. Also, this method doesn't seem to depend on the UI thread.